### PR TITLE
Enabling darwin universal x86_64/arm64 builds

### PR DESCRIPTION
**Pull request by brad-linden**

### CODE REVIEW

File: build-cmd.sh [hunk 1]
```diff
@@ -84,10 +84,15 @@ pushd "$ZLIB_SOURCE_DIR"
...
-            cc_opts="${TARGET_OPTS:--arch $AUTOBUILD_CONFIGURE_ARCH $LL_BUILD_RELEASE}"
-            ld_opts="-Wl,-install_name,\"${install_name}\" -Wl,-headerpad_max_install_names"
+            # todo - build support into autobuild for universal AUTOBUILD_CONFIGURE_ARCH
+            ARCH_ARGS="-arch x86_64 -arch arm64"
+            cc_opts="${TARGET_OPTS:-$ARCH_ARGS $LL_BUILD_RELEASE}"
+            ld_opts="$ARCH_ARGS -Wl,-install_name,\"${install_name}\" -Wl,-headerpad_max_install_names"
             export CC=clang
 
+            logical_cpus=$(sysctl hw.logicalcpu | cut -d ':' -d ' ' -f 2)
+            export MAKEFLAGS="-j${logical_cpus:-2}"
```
Comment (bennettgoble) [inline build-cmd.sh]: As of [v3.6.0](https://github.com/secondlife/autobuild/releases/tag/v3.6.0) you can use `AUTOBUILD_CPU_COUNT`, provided by the autobuild environment.

**Comment by bennettgoble** [comment]
Brad, can a build get kicked off with these changes? Actually, I'm a but unsure why one hasn't already as the build.yaml conditions are just `[push]`. I assume the problem may be that the default branch doesn't have actions.

File: build-cmd.sh (hunk 1 above)
Comment (brad-linden) [reply build-cmd.sh]: even better

**Comment by brad-linden** [comment]
merging in the actions branch seems to have done the trick for getting the build kicked off

**Comment by brad-linden** [comment]
looks like this won't be workable until we can use a runner with xcode-14.1 on it

**Comment by brad-linden** [comment]
huh, xcode-14.1 didn't help either

**Comment by brad-linden** [comment]
ok I figured out what was wrong.  it was succeeding by accident on my local arm64 mac, and as a result was building a bad library.  doing the 2 architecture builds in sequence and joining them with lipo is the only reliable way to get it working.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -1,10 +1,12 @@
 name: Build
 on: [push]
+env:
+  DEVELOPER_DIR: /Applications/Xcode_14.1.app/Contents/Developer
 jobs:
   build:
     strategy:
       matrix:
-        os: [windows-2022, macos-11, ubuntu-22.04]
+        os: [windows-2022, macos-12, ubuntu-22.04]
```
Comment (brad-linden) [inline .github/workflows/build.yaml]: this didn't end up being needed for fixing it, should I revert it?

Comment (bennettgoble) [reply .github/workflows/build.yaml]: Sure, let's try to keep all our builds on the same version.

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -1,10 +1,12 @@
 name: Build
 on: [push]
+env:
+  DEVELOPER_DIR: /Applications/Xcode_14.1.app/Contents/Developer
```
Comment (nat-goodspeed) [inline .github/workflows/build.yaml]: Is this necessary? Could we revert to default Xcode.app?

File: .github/workflows/build.yaml (hunk 2 above)
Comment (nat-goodspeed) [reply .github/workflows/build.yaml]: (apparently hasn't yet happened)

File: build-cmd.sh [hunk 4]
```diff
@@ -93,32 +93,41 @@ pushd "$ZLIB_SOURCE_DIR"
...
-                make test
-
-                # And wipe it
-                rm -rf ../Resources
-            fi
-
-            make distclean
+            export MAKEFLAGS="-j${AUTOBUILD_CPU_COUNT:-2}"
+
+            # todo - build support into autobuild for universal AUTOBUILD_CONFIGURE_ARCH
```
Comment (nat-goodspeed) [inline build-cmd.sh]: What would that involve?

**Review by nat-goodspeed — COMMENTED** [review]
Curious whether we could revert `build.yaml` as suggested.

File: build-cmd.sh (hunk 4 above)
Comment (brad-linden) [reply build-cmd.sh]: so, unfortunately this probably isn't possible the way I had envisioned.  apple's toolchain now lets you transparently build fat objects/archives/dylibs/executables simply by passing multiple arch flags in the appropriate CFLAGS places etc.  unfortunately almost every real project has some logic in its build system for selecting appropriate architecture specific source files and preprocessor defines, which conflict with this method.

anyways, if the new transparent method had worked, I would have wanted to build support for it into autobuild possibly by extending AUTOBUILD_CONFIGURE_ARCH to support multiple values being passed through, but seems like that won't be useful

**Comment by bennettgoble** [comment]
I don't see a macOS arm64 package in the [resulting artifacts](https://github.com/secondlife/3p-zlib-ng/actions/runs/3914331283). Is autobuild bottling arm64 incorrectly because AUTOBUILD_ARCHITECTURE=64?

**Comment by brad-linden** [comment]
this is correct. the resulting artifacts are universal in that they contain both arm64 and x86_64 object code for building and running on either architecture.  treating arm64 as a separate packaging platform would make it more difficult to produce universal executables

**Review by bennettgoble — APPROVED** [review]

**Comment by bennettgoble** [comment]
OK, great. Thought so, I was pondering whether we wanted a different package architecture string to express that it's universal: `zlib_ng-2.0.5.7f46b07-darwin-universal-7f46b07` or similar, but I think you're right; it'll be easier to just leave it as "64" (That's not wrong)

**Comment by brad-linden** [comment]
I've verified that those artifacts correctly contain fat libraries:

> brad@pixie:tmp {home} 0 🦄🌈 $ mv ~/Downloads/zlib_ng-2.0.5.7f46b07-darwin64-7f46b07.zip .
> brad@pixie:tmp {home} 0 🦄🌈 $ ls -l
> total 608
> -rw-r--r--@ 1 brad  staff  309394 Jan 13 15:05 zlib_ng-2.0.5.7f46b07-darwin64-7f46b07.zip
> brad@pixie:tmp {home} 0 🦄🌈 $ unzip zlib_ng-2.0.5.7f46b07-darwin64-7f46b07.zip
> Archive:  zlib_ng-2.0.5.7f46b07-darwin64-7f46b07.zip
> inflating: autobuild-graph.mermaid
> inflating: autobuild-results.json
> inflating: zlib_ng-2.0.5.7f46b07-darwin64-7f46b07.tar.zst
>  brad@pixie:tmp {home} 0 🦄🌈 $ tar -xvf zlib_ng-2.0.5.7f46b07-darwin64-7f46b07.tar.zst
> x include/zlib-ng/zlib.h
> x docs/zlib-ng/README.Linden
> x autobuild-package.xml
> x LICENSES/zlib-ng.txt
> x include/zlib-ng/zconf.h
> x lib/release/libz.a
>  brad@pixie:tmp {home} 0 🦄🌈 $ file lib/release/libz.a
> lib/release/libz.a: Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library [arm64:current ar archive random library]
> lib/release/libz.a (for architecture x86_64):	current ar archive random library
> lib/release/libz.a (for architecture arm64):	current ar archive random library
>  brad@pixie:tmp {home} 0 🦄🌈 $

**Comment by brad-linden** [comment]
> OK, great. Thought so, I was pondering whether we wanted a different package architecture string to express that it's universal: `zlib_ng-2.0.5.7f46b07-darwin-universal-7f46b07` or similar, but I think you're right; it'll be easier to just leave it as "64" (That's not wrong)

I believe we did do such a thing during the old ppc->intel migration and explicitly name things as universal, but I don't feel like we got a lot of value out of that, and I think we can streamline things for this time around

**Review by nat-goodspeed — APPROVED** [review]
LGTM

**Pull request merged**